README
WWW::UsePerl::Journal::Thread
=============================

DESCRIPTION

Using WWW::UsePerl::Journal to extract entry ids, each id can be used to seed
a thread of comments. Each comment can be accessed as an object, thus values
can be retrieve for each object variable.

DEPENDENCIES

The distribution requires the following modules:

  WWW::UsePerl::Journal => 0.15,
  LWP::UserAgent        => 0,
  HTTP::Request::Common => 0
  Time::Piece           => 0,

CHANGES

See the Changes file for a list of updates to the distribution.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

COPYRIGHT AND LICENSE

  Copyright (C) 2003-2007 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or 
  modify it under the same terms as Perl itself.

The full text of the licenses can be found in the Artistic and
COPYING files included with this module, or in perlartistic and
perlgpl in Perl 5.8.1 or later.



Changes
Revision history for Perl extension WWW::UsePerl::Journal::Thread.

0.14	2012-08-30
	- fixed test count on skip.

0.13	2012-08-26
	- use.perl.org is now read only. 
	  - thread id no longer supported.
        - removed License files from distro.
        - Artistic License v2 now sole license.
        - spelling fixes. (thanks to Nicholas Bamber, Florian Schlichting & 
          Gregor Herrmann).
        - added minimum perl version (5.006).
        - reworked Makefile.PL for clarity.
	- implemented Perl::Critic suggestions.
	- added META.json
	- change file dates changed to meet W3CDTF standards.

0.12	2010-09-30
	- fixed useperl.pl (v0.02) script in examples.
	- fixed META.yml - distribution name
	- this version was not released

0.11	2008-09-29
	- updated optional tests.
	- upgraded WWW-UsePerl-Journal required version to 0.22.
	- fixed regex to match comments for updated use.perl pages.
	- added a useperl.pl script to examples, which can extract all the 
	  journal posts and associated comments, for a given user.

0.10	2007-11-07
	- several fails pointed at bad error handling in the event of a webpage
	  not being returned. Further validation in the event of a page not
	  being found. Thanks to David Cantrell for the diagnostics.
	- bumped several prerequite versions in META.yml
	- incorporated internal debugging as per WUJ.

0.09	2007-10-31
	- relaxed date testing, as timezones can affect the date/time value
	  returned.
	- This is mostly a documentation and license update release.
	- Upped some of the versions of required modules.
	- New META.yml and associated tests added
	- Changed all instances of LICENCE to LICENSE.
	- Added LICENSE file.
	- Added an examples directory
	- POD/README updates.
	- UsePerl has changed its comment formatting.
	- simplified thread and comment storage respectively.
	- added error methods to access faults rather than die-ing.

0.08	2005-04-05
	- Updated test to reflect changes in use.perl
	- Added test scripts
	- Aligned version numbers

0.07	2004-04-22
	- Date in comment wasn't ... d'oh!
	- Date regex was balking. Fixed it up to be more economical and faster

0.06	2004-04-20
	- date/time testing removed due to odd behaviour with use.perl times

0.05	2004-04-19
	- Test::More added as a prerequisites for PPMs
	- Fixed Abstract in Makefile.PL

0.04	2003-11-10
	- Updated Test::Pod requirement
	- Between releases a few comments seem to have
	  gone back an hour on use.perl!!!

0.03	2003-10-08T13:04:18+01:00
	- META.yml added to distribution
	- POD updates

0.02	2003-08-14T12:23:02+01:00
	- POD updates and fixes  

0.01	2003-08-13T09:26:53+01:00
	- original version; based upon WWW::UsePerl::Journal 


t/lib/PingTest.pm
package PingTest;

use strict;
use warnings;

my $VERSION = '0.22';

#----------------------------------------------------------------------------

=head1 NAME

PingTest - private test module for check a network connection

=head1 DESCRIPTION

Provids one method, pingtest(), to check whether a network connection is
available.

=cut

# -------------------------------------
# Functions

# crude, but it'll hopefully do ;)
# ping returns 1 if unable to connect

sub pingtest {
    my $domain = shift || return 1;
    system("ping -q -c 1 $domain >/dev/null 2>&1");
    my $retcode = $? >> 8;
    return $retcode;
}

1;

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2009 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

The full text of the licenses can be found in the F<Artistic> and
F<COPYING> files included with this module, or in L<perlartistic> and
L<perlgpl> in Perl 5.8.1 or later.

=cut


examples/useperl.pl
#!/usr/bin/perl -w
use strict;
$|++;

my $VERSION = '0.02';

#----------------------------------------------------------------------------

=head1 NAME

useperl.pl - script to collect journal postings for a given user.

=head1 SYNOPSIS

  perl useperl.pl [--verbose] [--comments] [-y=<file>] [-h][-v] [-u=]<user>

=head1 DESCRIPTION

Given a user, set as either the last parameter, or with the command line option
-u, will retrieve all the journal postings made by that user on use.perl. In
addition the optional --comments flag includes any comments posted against that
journal entry.

If a --yaml file is given, the results are written to file in YAML format. If
the --verbose flag is given, the text is printed to STDOUT.

=cut

# -------------------------------------
# Library Modules

use File::Basename;
use File::Path;
use File::Slurp;
use Getopt::ArgvFile default=>1;
use Getopt::Long;
use WWW::UsePerl::Journal;
use WWW::UsePerl::Journal::Thread;
use YAML;

# -------------------------------------
# Variables

my (%options,@entries);
my $first = 1;

# -------------------------------------
# Program

##### INITIALISE #####

init_options();

my $journal = WWW::UsePerl::Journal->new($options{user});
#$journal->debug(1);
my @ids = $journal->entryids(ascending => 1);
for my $id (@ids) {
    my $entry = $journal->entry($id);

    my @comments;

    if($options{comments}) {
        my $thread = WWW::UsePerl::Journal::Thread->new(j => $journal, entry => $id);
        my @cids = $thread->commentids();
        for my $cid (@cids) {
            my $ccontent = $thread->comment($cid)->content();
            next    unless($ccontent);
            $ccontent =~ s!\t! !g;

            my $comment = {
                user    => $thread->comment($cid)->user,
                date    => $thread->comment($cid)->date->epoch,
                subject => $thread->comment($cid)->subject,
                content => $ccontent
            };

            push @comments, $comment;
        }
    }

    # clean content string
    my $content = $entry->content;
    $content =~ s!\t! !g;

    # save for later
    my $post = {
        id       => $id,
        date     => $entry->date->epoch,
        subject  => $entry->subject,
        content  => $content,
    };

    $post->{comments} = \@comments  if(@comments);
    push @entries, $post    if($options{yaml});

    if($options{verbose}) {
        # print for now
        print  "\n---- POST ----\n\n" unless($first-- > 0);
        printf "Link: http://use.perl.org/~barbie/%d\n", $post->{id};
        printf "Date: %s\n",    $post->{date};
        printf "Subject: %s\n", $post->{subject};
        printf "\n%s\n",        ($post->{content}||'');
        for my $comment (@{$post->{comments}}) {
            print "\n#### COMMENT ####\n\n";
            printf "User: %s\n",    $comment->{user};
            printf "Date: %s\n",    $comment->{date};
            printf "Subject: %s\n", $comment->{subject};
            printf "\n%s\n",        ($comment->{content}||'');
        }
    }
}


# save hash as YAML
write_file($options{yaml}, Dump(\@entries)) if($options{yaml});

#print "LOG: ".$journal->log()."\n";

# -------------------------------------
# Functions

sub init_options {
    GetOptions( \%options,
        'verbose',
        'comments',
        'yaml|y=s',
        'user|u=s',
        'help|h',
        'version|V'
    );

    _help(1) if($options{help});
    _help(0) if($options{version});

    if(defined $options{yaml} && ! -f $options{yaml}) {
        mkpath(dirname($options{yaml}));
    }

    $options{user} ||= $ARGV[0];
    if(!$options{user}) {
    	print "No user specified\n\n";
	    _help(1);
    }

    unless($options{yaml} || $options{verbose}) {
    	print "No output specified\n\n";
	    _help(1);
    }
}

sub _help {
    my $full = shift;

    if($full) {
        print <<HERE;

Usage: $0 \\
         [--yaml=<file>] [--verbose] [--comments] \\
         [-h] [-V] [--user=]<user>]

  --yaml=<file>     YAML output file
  --verbose         print output to STDOUT
  --comments        include comments from postings
  --user=<user>     named user
  -h                this help screen
  -V                program version

HERE

    }

    print "$0 v$VERSION\n\n";
    exit(0);
}

__END__

=back

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-UsePerl-Journal-Thread

=head1 SEE ALSO

L<WWW::UsePerl::Journal>

F<http://use.perl.org/>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2005-2008 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut



lib/WWW/UsePerl/Journal/Thread.pm
package WWW::UsePerl::Journal::Thread;

use strict;
use warnings;

use vars qw($VERSION);
$VERSION = '0.14';

#----------------------------------------------------------------------------

=head1 NAME

WWW::UsePerl::Journal::Thread - Handles the retrieval of UsePerl journal comment threads

=head1 SYNOPSIS

  use WWW::UsePerl::Journal;
  use WWW::UsePerl::Journal::Thread;

  my $journal = WWW::UsePerl::Journal->new('barbie');
  my @entries = $journal->entryids();

  my $thread = WWW::UsePerl::Journal::Thread->new(
        j       => $journal, 
        thread  => $entries[0]
  );

  my @comments = $thread->commentids();
  for my $id (@comments) {
    printf "\n----\n%s [%d %s %d] %s",
	  $thread->comment($id)->subject(),
	  $thread->comment($id)->score(),
	  $thread->comment($id)->user(),
	  $thread->comment($id)->uid(),
	  $thread->comment($id)->date(),
	  $thread->comment($id)->content();
  }

  my $threadid = $thread->thread();

=head1 DESCRIPTION

A collection of routines to handle the retrieval of threads from a
UsePerl (L<http://use.perl.org/>) journal entry.

Using WWW::UsePerl::Journal, journal entry ids can be obtain. Each entry id
can be used to obtain a comment thread. Each comment property is accessed
via a comment object from within the thread.

Note that as on late 2010 use.perl was decommissioned. A read-only version of
the site now exists on the perl.org servers, and a full database backup is
also available if you wish to host your own use.perl archive. 

A future edition of this distribution will allow a DBI interface to a local
database to retrieve journal entries.

=cut

# -------------------------------------
# Library Modules

use HTTP::Request::Common;
use LWP::UserAgent;
use Time::Piece;
use WWW::UsePerl::Journal::Comment;

# -------------------------------------
# Variables

use constant USEPERL => 'http://use.perl.org/use.perl.org';

my %months = (
	'January'   => 1,
	'February'  => 2,
	'March'     => 3,
	'April'     => 4,
	'May'       => 5,
	'June'      => 6,
	'July'      => 7,
	'August'    => 8,
	'September' => 9,
	'October'   => 10,
	'November'  => 11,
	'December'  => 12,
);

# -------------------------------------
# Public Interface

=head1 PUBLIC INTERFACE

=head2 The Constructor

=over 4

=item new

  use WWW::UsePerl::Journal;
  my $journal = WWW::UsePerl::Journal->new('barbie');

  use WWW::UsePerl::Journal::Thread;
  my $j = WWW::UsePerl::Journal::Thread->new(
            j       => $journal, 
            eid     => $entryid,
  );

Creates an thread instance for the specified journal entry. An entry ID 
returned from $journal->entryids() must use the entry => $entryid form to 
obtain the correct thread.

=back

=cut

sub new {
    my ($class,%opts) = @_;

    for(qw/j eid/) {
    	return	unless(exists $opts{$_});
    }

    die "No parent object"
	    unless $opts{j}->isa('WWW::UsePerl::Journal');

    my %atts = map {$_ => $opts{$_}} qw(j eid);
    my $self = bless \%atts, $class;

    return $self;
}

=head2 Methods

=over 4

=item thread()

Returns the current thread id.

=cut

sub thread {
    my $self = shift;
	$self->_commenthash	unless($self->{thread});
	return $self->{thread};
}

=item comment($commentid)

Returns a comment object of the given comment ID

=cut

sub comment {
    my $self = shift;
    my $cid  = shift;
	my %entries = $self->_commenthash;
    return $entries{$cid};
}

=item commentids()

Returns an ascending array of the comment IDs.

Can take an optional hash containing; {descending=>1} to return a descending
list of comment IDs, {ascending=>1} to return an ascending list or
{threaded=>1} to return a thread ordered list. 'ascending' being the default.

=back

=cut

sub commentids {
    my ($self,%hash) = @_;

	my ($key,$sorter) = ('_commentids_asc',\&_ascender);
   	   ($key,$sorter) = ('_commentids_dsc',\&_descender)	if(%hash && $hash{descending});
	   ($key,$sorter) = ('_commentids_thd',sub{-1})			if(%hash && $hash{threaded});

    $self->{$key} ||= do {
        my %entries = $self->_commenthash;
        my @ids = sort $sorter keys %entries;
        \@ids;
    };

    return @{$self->{$key}};
}

# -------------------------------------
# The Private Methods

# name: commenthash
# desc: Returns a hash of WWW::UsePerl::Journal::Comment objects

sub _commenthash {
    my $self = shift;

    return %{ $self->{_commenthash} }	if($self->{_commenthash});

	# URL depends upon which id we've been given, as thread and entry
	# are different, but both can still return the thread list, just in
	# different formats

    my $user = $self->{j}->user;
	my $url = USEPERL . "/_$user/journal/$self->{eid}.html";

	my $content;
    eval { $content = $self->{j}->{ua}->request(GET $url)->content; };
	return $self->{j}->error("could not create comment list") if($@ || !$content);

    if($self->{j}->debug) {
        $self->{j}->log('mess' => "#_commenthash: url=[$url]\n");
        $self->{j}->log('mess' => "#_commenthash: content=[$content]\n");
    }

	# main comment thread
	my %comments;
    my @cids = $content =~ m! <div\s+id="comment_top_(\d+)"  !sixg;
    if($self->{j}->debug) {
        $self->{j}->log('mess' => "#cids: @cids\n");
    }

    ($self->{thread}) = $content =~ m!sid=(\d+)!;

    for my $cid (@cids) {

        if($self->{j}->debug) {
            $self->{j}->log('mess' => "\n#_commenthash: cid=[$cid]\n");
        }

		next if($comments{$cid});

        my ($extract) = $content =~ m! (<ul\s+id="[^"]+"[^>]*>\s*<li\s+id="tree_$cid"\s+class="[^"]+">.*?<div\s+id="replyto_$cid"></div>) !six;
        if($self->{j}->debug) {
            $self->{j}->log('mess' => "\n#extract: [$extract]\n");
        }

		$comments{$cid} = WWW::UsePerl::Journal::Comment->new(
                j       => $self->{j},
                cid     => $cid,
                eid     => $self->{eid},
                extract => $extract
		);
	}

	%{ $self->{_commenthash} } = %comments;
	return %{ $self->{_commenthash} };
}

# sort methods

sub _ascender  { $a <=> $b }
sub _descender { $b <=> $a }

1;

__END__

=head1 SUPPORT

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please submit a bug to the RT system (see link below). However,
it would help greatly if you are able to pinpoint problems or even supply a
patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: L<http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-UsePerl-Journal-Thread>

=head1 SEE ALSO

F<http://use.perl.org/use.perl.org>

L<WWW::UsePerl::Journal>,
L<WWW::UsePerl::Journal::Server>

=head1 AUTHOR

Barbie, E<lt>barbie@cpan.orgE<gt>
for Miss Barbell Productions L<http://www.missbarbell.co.uk>.

=head1 CREDITS

Russell Matbouli, for creating WWW::UsePerl::Journal in the first place
and giving me the idea to extend it further.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2003-2012 Barbie for Miss Barbell Productions

This module is free software; you can redistribute it and/or
modify it under the Artistic Licence v2.

=cut


lib/WWW/UsePerl/Journal/Comment.pm
package WWW::UsePerl::Journal::Comment;

use strict;
use warnings;

use vars qw($VERSION $AUTOLOAD);
$VERSION = '0.14';

#----------------------------------------------------------------------------

=head1 NAME

WWW::UsePerl::Journal::Comment - Handles the retrieval of UsePerl journal entry comments.

=head1 SYNOPSIS

  my $comment = WWW::UsePerl::Journal::Comment->new(
      # required
      j       => $journal,
      cid     => $commentid,
      eid     => $entryid,
      extract => $extract
  );

  $comment->subject();

  # called from WWW::UsePerl::Journal::Thread object
  $thread->comment( $cid )->content();

=head1 DESCRIPTION

A collection of routines to handle the retrieval of comments from a
UsePerl (L<http://use.perl.org/>) journal entry.

=cut

# -------------------------------------
# Library Modules

use base qw(Class::Accessor::Fast);

use HTTP::Request::Common;
use LWP::UserAgent;
use Time::Piece;
use WWW::UsePerl::Journal;

# -------------------------------------
# Constants & Variables

use constant USEPERL => 'http://use.perl.org/use.perl.org';
use overload q{""}  => sub { $_[0]->stringify() };

my %months = (
	'January'   => 1,
	'February'  => 2,
	'March'     => 3,
	'April'     => 4,
	'May'       => 5,
	'June'      => 6,
	'July'      => 7,
	'August'    => 8,
	'September' => 9,
	'October'   => 10,
	'November'  => 11,
	'December'  => 12,
);

# -------------------------------------
# Public Interface

=head1 PUBLIC INTERFACE

=head2 The Constructor

=over 4

=item new

Each comment is retrieved as an object. Note that the parent object
(from WWW::UsePerl::Journal), thread id and comment id are mandatory
requirements to create the object.

=back

=cut

sub new {
    my ($class,%opts) = @_;

    for(qw/j eid cid extract/) {
    	return	unless(exists $opts{$_});
    }

    die "No parent object"
	    unless $opts{j}->isa('WWW::UsePerl::Journal');

    my %atts = map {$_ => $opts{$_}} qw(j eid cid extract);
    my $self = bless \%atts, $class;

    $self->_get_content();

    return $self;
}

#----------------------------------------------------------------------------
# Accessors

=head2 The Accessors

The following accessor methods are available:

  id
  date
  subject
  user
  uid
  score
  content

All functions can be called to return the current value of the associated
object variable.

=cut

__PACKAGE__->mk_accessors($_) for qw(cid date subject user uid score content);

=head2 Methods

=over 4

=item stringify - For String Context

When an object is called directly, stringification occurs. Safer to
use -E<gt>content instead.

=back

=cut

sub stringify {
    my $self = shift;
    return $self->content();
}

# -------------------------------------
# The Private Subs

# name:	_get_content
# args:	self .... object itself
# retv: content text
# desc: Retrieves the content and additional information for a given
#       comment. Splits the fields into object variables and returns
#       the content text

sub _get_content {
    my $self    = shift;

    my $content = $self->{extract};

    if($self->{j}->debug) {
        $self->{j}->log('mess' => "\n#_get_content: content=[$content]\n");
    }

    return $self->{j}->error("Error getting entry") unless $content;

    # remember there are different presentations for dates!!!!

	my ($string,$format);
	$content =~ s/\n//g;
	my @fields = ( $content =~ m!
            <li\s+id="tree_(\d+)"\s+class="comment[^"]*">   # comment id
    .*?     <h4><a[^>]+>([^<]+)</a>                         # subject
    .*?     <span\s+id="comment_score_\1"\s+class="score">
    .*?     Score:(\d+).*?</h4>                             # score
	.*?		<a\s+href="[./\w]+/index.html">\s*(\w+)         # username
	.*?		\((\d+)\)</a>						            # userid
            (?:\s+<span\s+class="otherdetails"
	.*?		    on\s+(\d+\.\d+.\d+\s\d+:\d+)  		        # date/time - "2003.05.20 17:31" or "Friday August 08 2003, @01:51PM"
    .*?     </span>)?
    .*?     <div\s+id="comment_body_\1">(.*?)</div>         # text
        !mixs );

    ($self->{pid}) = $content =~ m/id="commtree_(\d+)"/;
    
    if($self->{j}->debug) {
        $self->{j}->log('mess' => "\n#_get_content: fields=[".(join("][",map {$_||''} @fields))."]\n");
    }

    return  unless(@fields);

    if($fields[5]) {
        my ($year, $month, $day, $hr, $mi) = $fields[5] =~ m! (\d+)\.(\d+)\.(\d+) .*? (\d+):(\d+) !smx;
        unless($day) {
            my $amp;
            ($month, $day, $year, $hr, $mi, $amp) = $fields[5] =~ m! \w+\s+ (\w+) \s+(\d+)\s*(\d*), \s+ @(\d+):(\d+)([AP]M) !smx;
            $month = $months{$month};
            $year = (localtime)[5]  unless($year);	# current year formatting drops the year.
            $hr += 12 if ($amp eq 'PM');
            $hr = 0 if $hr == 24;
        }

        if($self->{j}->debug) {
            $self->{j}->log('mess' => "\n#_get_content: date=[$year $month $day ${hr}:$mi]\n");
        }

        # sometimes Time::Piece can't parse the date :(
        eval {
            $self->{date} = Time::Piece->strptime(
                "$month $day $year ${hr}:$mi",
                '%m %d %Y %H:%M'
            );
        };

        if($self->{j}->debug) {
            $self->{j}->log('mess' => "\n#_get_content: date=[".$self->{date}."]\n");
        }
    }

	$self->{subject}	= $fields[1];
	$self->{score}		= $fields[2];
	$self->{user}		= $fields[3];
	$self->{uid}		= $fields[4];
	$self->{content}	= $fields[6];

	return  unless($self->{content});				# What no content!

	$self->{content} =~ s!(\s+<(?:p|br /)>)*$!!gi;	# remove trailing whitespace formatting
	$self->{content} =~ s!\s+(<(p|br /)>)!$1!gi;	# remove whitespace before whitespace formatting
	$self->{content} =~ s!(<(p|br /)>){2,}!<p>!gi;	# remove repeated whitespace formatting

    return;
}

sub DESTROY {}

1;

__END__

=head1 SUPPORT

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please submit a bug to the RT system (see link below). However,
it would help greatly if you are able to pinpoint problems or even supply a
patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me by sending an email
to barbie@cpan.org .

RT: L<http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-UsePerl-Journal-Thread>

=head1 SEE ALSO

F<http://use.perl.org/use.perl.org>

L<WWW::UsePerl::Journal>,
L<WWW::UsePerl::Journal::Server>

=head1 CREDITS

Russell Matbouli, for creating L<WWW::UsePerl::Journal> in the first place
and giving me the idea to extend it further.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2003-2012 Barbie for Miss Barbell Productions.

This module is free software; you can redistribute it and/or
modify it under the Artistic Licence v2.

=cut
